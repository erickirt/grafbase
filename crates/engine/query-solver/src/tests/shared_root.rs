use crate::assert_solving_snapshots;

const SCHEMA: &str = r###"
enum join__Graph {
  CATEGORY @join__graph(name: "category", url: "http://localhost:4200/shared-root/category")
  NAME @join__graph(name: "name", url: "http://localhost:4200/shared-root/name")
  PRICE @join__graph(name: "price", url: "http://localhost:4200/shared-root/price")
}

type Product
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
  nested: Nested!
  category: String @join__field(graph: CATEGORY)
  name: String @join__field(graph: NAME)
  price: Float @join__field(graph: PRICE)
}

type Nested
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
}

type Query
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  product: Product
  products: [Product]
}
"###;

#[tokio::test]
async fn all_fields() {
    assert_solving_snapshots!(
        "all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            id
            name
          }
        }
        "#
    );
}

#[tokio::test]
async fn nested_id_field_with_all_fields() {
    assert_solving_snapshots!(
        "nested_id_with_all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            name
            nested { id }
          }
        }
        "#
    );
}

#[tokio::test]
async fn single_field() {
    assert_solving_snapshots!(
        "single_field",
        SCHEMA,
        r#"
        query {
          products {
            price
          }
        }
        "#
    );
}
