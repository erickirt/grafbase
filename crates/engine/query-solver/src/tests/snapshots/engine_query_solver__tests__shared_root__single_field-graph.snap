---
source: crates/engine/query-solver/src/tests/shared_root.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Query.products\" ]\n    2 [ label = \"Product.price\" ]\n    7 [ label = \"Root#price\", shape=parallelogram, color=dodgerblue ]\n    8 [ label = \"products#price\", shape=box, color=dodgerblue ]\n    9 [ label = \"price#price\", shape=box, color=dodgerblue ]\n    0 -> 1 [ label = \"\" ]\n    1 -> 2 [ label = \"\" ]\n    0 -> 7 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    0 -> 7 [ label = \"\", style=dashed,arrowhead=none ]\n    7 -> 8 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    8 -> 1 [ label = \"\", color=violet,arrowhead=none ]\n    8 -> 9 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    9 -> 2 [ label = \"\", color=violet,arrowhead=none ]\n}\n"
---
digraph {
    0 [ root]
    1 [ Query.products]
    2 [ Product.price]
    7 [ Root#price]
    8 [ products#price]
    9 [ price#price]
    0 -> 1 [ label = "Field" ]
    1 -> 2 [ label = "Field" ]
    0 -> 7 [ label = "CreateChildResolver" ]
    0 -> 7 [ label = "HasChildResolver" ]
    7 -> 8 [ label = "CanProvide" ]
    8 -> 1 [ label = "Provides" ]
    8 -> 9 [ label = "CanProvide" ]
    9 -> 2 [ label = "Provides" ]
}
